File: sdk/kotlin/src/main/kotlin/com/lunziko/aiengine/LunzikoAIEngine.kt
/*
 * SDK client Kotlin de Lunziko AI Engine — autonome, indépendant de Platform.
 * JDK 11+ (java.net.http) + kotlinx.serialization.
 *
 *   val ai = LunzikoAIEngine("http://localhost:8770", apiKey = "...")
 *   val r = ai.chat(listOf(ChatMessage.user("Bonjour")))
 *   println(r["content"])
 */
package com.lunziko.aiengine

import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import kotlinx.serialization.json.*

/** Erreur renvoyée par le gateway. */
class AIEngineException(val status: Int, val detail: String) :
    RuntimeException("AI Engine $status: $detail")

/** Message de conversation. */
data class ChatMessage(val role: String, val content: String) {
    companion object {
        fun user(content: String) = ChatMessage("user", content)
        fun assistant(content: String) = ChatMessage("assistant", content)
    }
}

class LunzikoAIEngine(
    baseUrl: String,
    private val apiKey: String? = null,
    private val http: HttpClient = HttpClient.newHttpClient(),
) {
    private val baseUrl = baseUrl.trimEnd('/')
    private val json = Json { ignoreUnknownKeys = true }

    private fun request(method: String, path: String, body: JsonElement? = null): JsonObject {
        val builder = HttpRequest.newBuilder(URI.create("$baseUrl$path"))
            .header("content-type", "application/json")
        apiKey?.let { builder.header("X-API-Key", it) }
        val publisher = if (body == null) HttpRequest.BodyPublishers.noBody()
            else HttpRequest.BodyPublishers.ofString(body.toString())
        when (method) {
            "GET" -> builder.GET()
            "PUT" -> builder.PUT(publisher)
            "DELETE" -> builder.DELETE()
            else -> builder.POST(publisher)
        }
        val res = http.send(builder.build(), HttpResponse.BodyHandlers.ofString())
        if (res.statusCode() >= 400) throw AIEngineException(res.statusCode(), res.body())
        val txt = res.body()
        return if (txt.isBlank()) JsonObject(emptyMap()) else json.parseToJsonElement(txt).jsonObject
    }

    /** GET générique (n'importe quel endpoint). */
    fun get(path: String): JsonObject = request("GET", path)

    /** POST générique. */
    fun post(path: String, body: JsonObject = JsonObject(emptyMap())): JsonObject =
        request("POST", path, body)

    // --- Système ---
    fun health(): JsonObject = get("/health")
    fun providers(): JsonObject = get("/v1/providers")

    // --- LLM & embeddings ---
    fun chat(messages: List<ChatMessage>, provider: String? = null, system: String? = null,
             model: String? = null, maxTokens: Int? = null): JsonObject {
        val body = buildJsonObject {
            putJsonArray("messages") {
                messages.forEach { add(buildJsonObject { put("role", it.role); put("content", it.content) }) }
            }
            provider?.let { put("provider", it) }
            system?.let { put("system", it) }
            model?.let { put("model", it) }
            maxTokens?.let { put("max_tokens", it) }
        }
        return post("/v1/chat", body)
    }

    fun embed(texts: List<String>): JsonObject =
        post("/v1/embed", buildJsonObject { putJsonArray("texts") { texts.forEach { add(it) } } })

    // --- RAG ---
    fun ragSearch(namespace: String, query: String, k: Int = 5): JsonObject =
        post("/v1/rag/search", buildJsonObject { put("namespace", namespace); put("query", query); put("k", k) })
    fun ragQuery(namespace: String, query: String, k: Int = 5, provider: String? = null): JsonObject =
        post("/v1/rag/query", buildJsonObject {
            put("namespace", namespace); put("query", query); put("k", k); provider?.let { put("provider", it) }
        })

    // --- Mémoire & knowledge ---
    fun memorySave(userId: String, key: String, value: String, category: String = "general"): JsonObject =
        post("/v1/memory/save", buildJsonObject {
            put("user_id", userId); put("key", key); put("value", value); put("category", category)
        })
    fun memoryRecall(userId: String, query: String, k: Int = 5): JsonObject =
        post("/v1/memory/recall", buildJsonObject { put("user_id", userId); put("query", query); put("k", k) })
    fun knowledgeSearch(org: String, query: String, k: Int = 5): JsonObject =
        post("/v1/knowledge/search", buildJsonObject { put("org", org); put("query", query); put("k", k) })

    // --- Agent & outils (A-4b) ---
    fun agent(query: String, userId: String? = null, org: String? = null, provider: String? = null): JsonObject =
        post("/v1/agent/run", buildJsonObject {
            put("query", query); userId?.let { put("user_id", it) }; org?.let { put("org", it) }
            provider?.let { put("provider", it) }
        })
    fun act(query: String, tools: List<String>? = null, provider: String? = null): JsonObject =
        post("/v1/agent/act", buildJsonObject {
            put("query", query)
            tools?.let { putJsonArray("tools") { it.forEach { t -> add(t) } } }
            provider?.let { put("provider", it) }
        })
    fun tools(): JsonObject = get("/v1/tools")
    fun runTool(name: String, arguments: JsonObject): JsonObject =
        post("/v1/tools/run", buildJsonObject { put("name", name); put("arguments", arguments) })

    // --- Écosystème ---
    fun ecosystemApps(): JsonObject = get("/v1/ecosystem/apps")
    fun ecosystemSearch(query: String, k: Int = 5): JsonObject =
        post("/v1/ecosystem/search", buildJsonObject { put("query", query); put("k", k) })

    // --- Activité & contexte ---
    fun activityLog(userId: String, app: String, action: String, target: String = "",
                    status: String = "ok", detail: String = ""): JsonObject =
        post("/v1/activity/log", buildJsonObject {
            put("user_id", userId); put("app", app); put("action", action)
            put("target", target); put("status", status); put("detail", detail)
        })
    fun contextAssemble(userId: String, query: String = "", app: String? = null): JsonObject =
        post("/v1/context/assemble", buildJsonObject {
            put("user_id", userId); put("query", query); app?.let { put("app", it) }
        })

    // --- Assistant scopé & handoff ---
    fun assistantAsk(app: String, query: String, userId: String? = null): JsonObject =
        post("/v1/assistant/$app/ask", buildJsonObject { put("query", query); userId?.let { put("user_id", it) } })
    fun assistantScope(app: String): JsonObject = get("/v1/assistant/$app/scope")
    fun handoffOpenWith(fromApp: String, filename: String): JsonObject =
        post("/v1/handoff/open-with", buildJsonObject { put("from_app", fromApp); put("filename", filename) })

    // --- Neural & données ---
    fun neuralRoute(query: String): JsonObject =
        post("/v1/neural/route", buildJsonObject { put("query", query) })
    fun mlPredict(name: String, text: String): JsonObject =
        post("/v1/neural/ml/predict", buildJsonObject { put("name", name); put("text", text) })
    fun dataCleanText(texts: List<String>, minLen: Int = 1): JsonObject =
        post("/v1/data/clean-text", buildJsonObject {
            putJsonArray("texts") { texts.forEach { add(it) } }; put("min_len", minLen)
        })

    // --- Automatisation (A-10) ---
    fun runFlow(name: String, input: JsonObject): JsonObject =
        post("/v1/automation/flows/$name/run", buildJsonObject { put("input", input) })
}

File: sdk/dart/lib/lunziko_ai_engine.dart
/// SDK client de Lunziko AI Engine — autonome, indépendant de Platform.
///
/// Un seul client pour tout le gateway `/v1/*`.
///
///     final ai = LunzikoAIEngine(baseUrl: 'http://localhost:8770', apiKey: '...');
///     final r = await ai.chat([ChatMessage.user('Bonjour')]);
///     print(r['content']);
library lunziko_ai_engine;

import 'dart:convert';
import 'package:http/http.dart' as http;

/// Message de conversation.
class ChatMessage {
  final String role;
  final String content;
  ChatMessage(this.role, this.content);
  factory ChatMessage.user(String content) => ChatMessage('user', content);
  factory ChatMessage.assistant(String content) => ChatMessage('assistant', content);
  Map<String, dynamic> toJson() => {'role': role, 'content': content};
}

/// Erreur renvoyée par le gateway.
class AIEngineException implements Exception {
  final int status;
  final String detail;
  AIEngineException(this.status, this.detail);
  @override
  String toString() => 'AI Engine $status: $detail';
}

class LunzikoAIEngine {
  final String baseUrl;
  final String? apiKey;
  final http.Client _client;

  LunzikoAIEngine({required String baseUrl, this.apiKey, http.Client? client})
      : baseUrl = baseUrl.replaceAll(RegExp(r'/+$'), ''),
        _client = client ?? http.Client();

  Future<dynamic> _request(String method, String path, [Object? body]) async {
    final headers = {'content-type': 'application/json'};
    if (apiKey != null) headers['X-API-Key'] = apiKey!;
    final uri = Uri.parse('$baseUrl$path');
    final encoded = body == null ? null : jsonEncode(body);
    late http.Response res;
    switch (method) {
      case 'GET':
        res = await _client.get(uri, headers: headers);
        break;
      case 'PUT':
        res = await _client.put(uri, headers: headers, body: encoded);
        break;
      case 'DELETE':
        res = await _client.delete(uri, headers: headers);
        break;
      default:
        res = await _client.post(uri, headers: headers, body: encoded);
    }
    if (res.statusCode >= 400) {
      throw AIEngineException(res.statusCode, res.body);
    }
    return res.body.isEmpty ? null : jsonDecode(res.body);
  }

  /// GET générique (n'importe quel endpoint du gateway).
  Future<Map<String, dynamic>> get(String path) async =>
      (await _request('GET', path)) as Map<String, dynamic>;

  /// POST générique.
  Future<Map<String, dynamic>> post(String path, [Object body = const {}]) async =>
      (await _request('POST', path, body)) as Map<String, dynamic>;

  void close() => _client.close();

  // --- Système ---
  Future<Map<String, dynamic>> health() => get('/health');
  Future<Map<String, dynamic>> providers() => get('/v1/providers');

  // --- LLM & embeddings ---
  Future<Map<String, dynamic>> chat(List<ChatMessage> messages,
          {String? provider, String? system, String? model, int? maxTokens}) =>
      post('/v1/chat', {
        'messages': messages.map((m) => m.toJson()).toList(),
        if (provider != null) 'provider': provider,
        if (system != null) 'system': system,
        if (model != null) 'model': model,
        if (maxTokens != null) 'max_tokens': maxTokens,
      });

  Future<Map<String, dynamic>> embed(List<String> texts) => post('/v1/embed', {'texts': texts});

  // --- RAG ---
  Future<Map<String, dynamic>> ragIndex(String namespace, String id, String text,
          {Map<String, dynamic>? meta}) =>
      post('/v1/rag/index', {'namespace': namespace, 'id': id, 'text': text, 'meta': meta ?? {}});
  Future<Map<String, dynamic>> ragSearch(String namespace, String query, {int k = 5}) =>
      post('/v1/rag/search', {'namespace': namespace, 'query': query, 'k': k});
  Future<Map<String, dynamic>> ragQuery(String namespace, String query,
          {int k = 5, String? provider}) =>
      post('/v1/rag/query',
          {'namespace': namespace, 'query': query, 'k': k, if (provider != null) 'provider': provider});

  // --- Mémoire & knowledge ---
  Future<Map<String, dynamic>> memorySave(String userId, String key, String value,
          {String category = 'general'}) =>
      post('/v1/memory/save', {'user_id': userId, 'key': key, 'value': value, 'category': category});
  Future<Map<String, dynamic>> memoryRecall(String userId, String query, {int k = 5}) =>
      post('/v1/memory/recall', {'user_id': userId, 'query': query, 'k': k});
  Future<Map<String, dynamic>> knowledgeSearch(String org, String query, {int k = 5}) =>
      post('/v1/knowledge/search', {'org': org, 'query': query, 'k': k});

  // --- Agent & outils (A-4b) ---
  Future<Map<String, dynamic>> agent(String query, {Map<String, dynamic>? options}) =>
      post('/v1/agent/run', {'query': query, ...?options});
  Future<Map<String, dynamic>> act(String query, {List<String>? tools, String? provider}) =>
      post('/v1/agent/act',
          {'query': query, if (tools != null) 'tools': tools, if (provider != null) 'provider': provider});
  Future<Map<String, dynamic>> tools() => get('/v1/tools');
  Future<Map<String, dynamic>> runTool(String name, Map<String, dynamic> arguments) =>
      post('/v1/tools/run', {'name': name, 'arguments': arguments});

  // --- Écosystème ---
  Future<Map<String, dynamic>> ecosystemApps() => get('/v1/ecosystem/apps');
  Future<Map<String, dynamic>> ecosystemSearch(String query, {int k = 5}) =>
      post('/v1/ecosystem/search', {'query': query, 'k': k});

  // --- Activité & contexte ---
  Future<Map<String, dynamic>> activityLog(String userId, String app, String action,
          {String target = '', String status = 'ok', String detail = ''}) =>
      post('/v1/activity/log', {
        'user_id': userId, 'app': app, 'action': action,
        'target': target, 'status': status, 'detail': detail,
      });
  Future<Map<String, dynamic>> contextAssemble(String userId, {String query = '', String? app}) =>
      post('/v1/context/assemble',
          {'user_id': userId, 'query': query, if (app != null) 'app': app});

  // --- Assistant scopé & handoff ---
  Future<Map<String, dynamic>> assistantAsk(String app, String query, {String? userId}) =>
      post('/v1/assistant/$app/ask', {'query': query, if (userId != null) 'user_id': userId});
  Future<Map<String, dynamic>> assistantScope(String app) => get('/v1/assistant/$app/scope');
  Future<Map<String, dynamic>> handoffOpenWith(String fromApp, String filename) =>
      post('/v1/handoff/open-with', {'from_app': fromApp, 'filename': filename});

  // --- Neural (routeur + ML) & données ---
  Future<Map<String, dynamic>> neuralRoute(String query) => post('/v1/neural/route', {'query': query});
  Future<Map<String, dynamic>> mlPredict(String name, String text) =>
      post('/v1/neural/ml/predict', {'name': name, 'text': text});
  Future<Map<String, dynamic>> dataCleanText(List<String> texts, {int minLen = 1}) =>
      post('/v1/data/clean-text', {'texts': texts, 'min_len': minLen});

  // --- Automatisation (A-10) ---
  Future<Map<String, dynamic>> runFlow(String name, Map<String, dynamic> input) =>
      post('/v1/automation/flows/$name/run', {'input': input});
}
